lib/c-api/src/wasm_c_api/wasi/mod.rs

```diff
@@ -10,20 +10,517 @@ use super::{
+        let result = unsafe {
+            (self_write)(
+                data.as_mut_ptr() as *const c_void,
+                buf.as_ptr() as *const c_char,
+                buf.len(),
+                false,
+            )
+        };
+        if result >= 0 {
+            Ok(result.try_into().unwrap_or(0))
+        } else {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                format!(
+                    "could not write {} bytes to wasi_pipe_t: {result}",
+                    buf.len()
+                ),
+            ))
+        }
+    }
+    fn flush(&mut self) -> io::Result<()> {
+        let self_write = self.write;
+        let mut data = self.get_data_mut("flush")?;
+        let bytes_to_write = &[];
+        let result: i64 = unsafe {
+            (self_write)(
+                data.as_mut_ptr() as *const c_void,
+                bytes_to_write.as_ptr(),
+                0,
+                true,
+            )
+        };
+        if result >= 0 {
+            Ok(())
+        } else {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                format!("could not flush wasi_pipe_t: {result}"),
+            ))
+        }
+    }
+}
+
+impl io::Seek for wasi_pipe_t {
+    fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
+        let self_seek = self.seek;
+        let mut data = self.get_data_mut("seek")?;
+        let (id, pos) = match pos {
+            SeekFrom::Start(s) => (0, s as i64),
+            SeekFrom::End(s) => (1, s),
+            SeekFrom::Current(s) => (2, s),
+        };
+        let result = unsafe { (self_seek)(data.as_mut_ptr() as *const c_void, id, pos) };
+        if result >= 0 {
+            Ok(result.try_into().unwrap_or(0))
+        } else {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                format!("could not seek to {pos:?} wasi_pipe_t: {result}"),
+            ))
+        }
+    }
+}
+
+impl VirtualFile for wasi_pipe_t {
+    fn last_accessed(&self) -> u64 {
+        0
+    }
+    fn last_modified(&self) -> u64 {
+        0
+    }
+    fn created_time(&self) -> u64 {
+        0
+    }
+    fn size(&self) -> u64 {
+        self.get_data_mut("size")
+            .map(|s| s.len() as u64)
+            .unwrap_or(0)
+    }
+    fn set_len(&mut self, _: u64) -> Result<(), FsError> {
+        Ok(())
+    }
+    fn unlink(&mut self) -> Result<(), FsError> {
+        Ok(())
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_pipe_new_internal(
+    read: WasiConsoleIoReadCallback,
+    write: WasiConsoleIoWriteCallback,
+    seek: WasiConsoleIoSeekCallback,
+    destructor: WasiConsoleIoEnvDestructor,
+    env_data: *const c_void,
+    env_data_len: usize,
+) -> *mut wasi_pipe_t {
+    let data_vec: Vec<c_char> =
+        std::slice::from_raw_parts(env_data as *const c_char, env_data_len).to_vec();
+
+    Box::leak(Box::new(wasi_pipe_t {
```

Comment:
`Box::leak` internally just uses `ManuallyDrop` itself, I'm keeping this as is, because it's easier to read `Box::leak` -> `Box::from_raw`, no extra imports needed.


https://doc.rust-lang.org/src/alloc/boxed.rs.html#1200